webapp/sources/ldap-inventory/inventory-repository/src/main/scala/com/normation/inventory/ldap/core/InventoryMapper.scala

```diff
@@ -1065,30 +1065,33 @@ class InventoryMapper(
       hostname       <- entry.required(A_HOSTNAME).toIO
       rootUser       <- entry.required(A_ROOT_USER).toIO
       policyServerId <- entry.required(A_POLICY_SERVER_UUID).toIO
-      publicKeys      = entry.valuesFor(A_PKEYS).map(Some(_))
       agentNames     <- {
-        val agents        = entry.valuesFor(A_AGENTS_NAME).toSeq.map(Some(_))
-        val agentWithKeys = agents.zipAll(publicKeys, None, None).filter(_._1.isDefined)
-        ZIO
-          .foreach(agentWithKeys) {
-            case (opt, key) =>
-              ((opt, key) match {
-                case (Some(agent), key) =>
-                  AgentInfoSerialisation
-                    .parseJson(agent, key)
-                    .chainError(s"Error when parsing agent security token '${agent}'")
-                case (None, _)          =>
-                  InventoryMappingRudderError.MissingMandatory("Error when parsing agent security token: agent is undefined").fail
-              }).foldZIO(
-                err =>
-                  InventoryDataLogger.error(
-                    s"Error when parsing agent information for node '${id.value}': that agent will be " +
-                    s"ignored for the node, which will likely cause problem like the node being ignored: ${err.fullMsg}"
-                  ) *> None.succeed,
-                ok => Some(ok).succeed
-              )
+        val agents = entry.valuesFor(A_AGENTS_NAME).toList
+        (ZIO
+          .foreach(agents) {
+            case agent =>
+              AgentInfoSerialisation
+                .parseJson(agent)
+                .chainError(s"Error when parsing agent security token '${agent}'")
+          })
+          .map(_.distinct)
+          .foldZIO(
+            err =>
+              InventoryDataLogger.error(
+                s"Error when parsing agent information for node '${id.value}': that agent will be " +
+                s"ignored for the node, which will likely cause problem like the node being ignored: ${err.fullMsg}"
+              ) *> None.succeed,
+            ok => Some(ok).succeed
+          )
+          .flatMap {
+            // we must have exactly one agent
+            case Some(a :: Nil) => List(a).succeed
+            case Some(several)  =>
+              InventoryDataLogger.error(
+                s"Error: inventory for node '${}' has several (${several.size}) different agents defined, which is not supported. Ignoring all agents."
+              ) *> Nil.succeed
+            case None           => Nil.succeed
```

Comment:
Okay there is some additional checks, but the whole logic could all be under the body of the `foldZIO` ? we can easily miss the `map(_.distinct)` (I'm also curious why it's necessary) and the whole process when it is split up like that

Reply:
I think it's purelly historical aggregation, I don't see specific reasons for not merging everything. 
The distinct is because once upon a time, you could have several agent. Then only one, but you didn't want to have a duplicate xml agent tag leading to a false error. I'm not sure we can still have duplicate xml part, but in doubts...